File: Code Samples/Linked List - Harsh Duche/PYTHON/DoublyLL.py

class DoublyNode:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    # Create operation: Adding a new node to the end
    def add_node(self, data):
        new_node = DoublyNode(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
            new_node.prev = current

    # Read operation: Accessing elements by index
    def get_node(self, index):
        current = self.head
        for i in range(index):
            if current is None:
                return -1  # Index out of bounds
            current = current.next
        return current.data if current else -1  # Node not found

    # Update operation: Modifying an element by index
    def update_node(self, index, data):
        current = self.head
        for i in range(index):
            if current is None:
                return
            current = current.next
        if current:
            current.data = data

    # Delete operation: Removing a node by index (without additional data structures)
    def delete_node(self, index):
        if index == 0:
            self.head = self.head.next
            if self.head:
                self.head.prev = None
        else:
            current = self.head
            for i in range(index - 1):
                if current is None:
                    return
                current = current.next
            if current and current.next:
                current.next = current.next.next
                if current.next:
                    current.next.prev = current

    # isEmpty operation
    def is_empty(self):
        return self.head is None

# Example usage
doubly_linked_list = DoublyLinkedList()

# Create operation
doubly_linked_list.add_node(1)
doubly_linked_list.add_node(2)
doubly_linked_list.add_node(3)

# Read operation
print("Element at index 1:", doubly_linked_list.get_node(1))

# Update operation
doubly_linked_list.update_node(1, 4)

# Delete operation
doubly_linked_list.delete_node(1)

# isEmpty operation
print("Is the list empty?", doubly_linked_list.is_empty())

File: Code Samples/Linked List - Harsh Duche/Java/DoublyLL.java
class DoublyNode {
    int data;
    DoublyNode next;
    DoublyNode prev;

    DoublyNode(int data) {
        this.data = data;
    }
}

class DoublyLinkedList {
    DoublyNode head;

    // Create operation: Adding a new node to the end
    public void addNode(int data) {
        DoublyNode newNode = new DoublyNode(data);
        if (head == null) {
            head = newNode;
        } else {
            DoublyNode current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
            newNode.prev = current;
        }
    }

    // Read operation: Accessing elements by index
    public int getNode(int index) {
        DoublyNode current = head;
        for (int i = 0; i < index && current != null; i++) {
            current = current.next;
        }
        if (current != null) {
            return current.data;
        } else {
            return -1; // Index out of bounds
        }
    }

    // Update operation: Modifying an element by index
    public void updateNode(int index, int data) {
        DoublyNode current = head;
        for (int i = 0; i < index && current != null; i++) {
            current = current.next;
        }
        if (current != null) {
            current.data = data;
        }
    }

    // Delete operation: Removing a node by index (without additional data structures)
    public void deleteNode(int index) {
        if (index == 0) {
            head = head.next;
            head.prev = null;
        } else {
            DoublyNode current = head;
            for (int i = 0; i < index - 1 && current != null; i++) {
                current = current.next;
            }
            if (current != null && current.next != null) {
                current.next = current.next.next;
                if (current.next != null) {
                    current.next.prev = current;
                }
            }
        }
    }

    // isEmpty operation
    public boolean isEmpty() {
        return head == null;
    }
}

public class DoublyLL {
    public static void main(String[] args) {
        DoublyLinkedList list = new DoublyLinkedList();

        // Create operation
        list.addNode(1);
        list.addNode(2);
        list.addNode(3);

        // Read operation
        System.out.println("Element at index 1: " + list.getNode(1));

        // Update operation
        list.updateNode(1, 4);

        // Delete operation
        list.deleteNode(1);

        // isEmpty operation
        System.out.println("Is the list empty? " + list.isEmpty());
    }
}
